ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -81,7 +89,47 @@ let ntlm_auth uname passwd : (unit, exn) result =
     Ok ()
   with _ -> Error (auth_ex uname)
 
+let get_domain_info_from_db () =
+  Server_helpers.exec_with_new_task "retrieving external auth domain workgroup"
+  @@ fun __context ->
+  let host = Helpers.get_localhost ~__context in
+  let service_name =
+    Db.Host.get_external_auth_service_name ~__context ~self:host
+  in
+  let workgroup, netbios_name =
+    Db.Host.get_external_auth_configuration ~__context ~self:host
+    |> fun config ->
+    (List.assoc_opt "workgroup" config, List.assoc_opt "netbios_name" config)
+  in
+  {service_name; workgroup; netbios_name}
+
 module Ldap = struct
+  (*
+   * Escape characters according to
+   * https://docs.microsoft.com/en-gb/windows/win32/adsi/search-filter-syntax?redirectedfrom=MSDN#special-characters
+   * *)
+  let escape_map =
+    [
+      ('*', "\\2a")
+    ; ('(', "\\28")
+    ; (')', "\\29")
+    ; ('\\', "\\5d")
+    ; ('\000', "\\00")
+    ; ('/', "\\2f")
+    ]
+
+  let escape str =
+    String.to_seq str
+    |> List.of_seq
+    |> List.map (fun k ->
+           match List.assoc_opt k escape_map with
+           | Some v ->
+               v
+           | None ->
+               String.make 1 k
```

Comment:
Creating lots of small 1-byte strings and then concatenating them together is not very efficient.
We can actually use sequences more efficiently here:
There is a `Seq.map`, and it is more efficient to apply this prior to conversion to list (sequences are traversed once and Seq.map operates on one element at a time, whereas List.map creates intermediate lists). 
And because we sometimes want to keep characters as they are (replace 1 char with 1 char), and sometimes expand them (escaping grows the string) then we can use Seq.flat_map for this purpose. So something like:
```ocaml
let escape_map = function
 | '*' -> String.to_seq "\\2a"
...
  | c -> Seq.return c

let escape str =
   str |> String.to_seq
  |> Seq.flat_map escape_map
  |> String.of_seq
```

Reply:
(Although in this case this is not performance-critical code, the faster version is actually shorter than the original, if one wants to optimize this further could use a Buffer and build the string that way, but I think the sequence is a nicer, more functional abstraction, so we can keep that).

Reply:
Good call. The fastest way to do this would be to scan for characters to be escaped - if none are in the string, which I would assume is the common case, just do nothing, otherwise do what Edvin suggests. The predicate `needs_escaping` would stop at the first characters that needs an escape or go to the end of the string otherwise and return false.

Reply:
Yes, this code is not performance critical,  ( I would say more than 99% cases will just get the subject ID from winbind cache, and thus, this code is not ran at all).
please se the updated code, which I think is more clear.